File: agentbench-backend-fixed/backend/app/schemas/benchmark.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict


class BenchmarkCreate(BaseModel):
    title: str
    description: str
    task: str


class BenchmarkUpdate(BaseModel):
    title: str
    description: str
    task: str


class BenchmarkResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str
    task: str
    created_by: int
    created_at: Optional[datetime] = None


class BenchmarkRunRequest(BaseModel):
    task_name: str
    prompt: str
    expected_output: Optional[str] = None
    providers: List[str]


class BenchmarkResult(BaseModel):
    provider: str
    # None when no active agent was configured for this provider.
    model: Optional[str] = None
    score: float = 0
    latency: float = 0
    tokens: int = 0
    cost: float = 0
    output: str = ""
    success: bool = False
    error: Optional[str] = None


class BenchmarkRunResponse(BaseModel):
    benchmark_id: int
    status: str
    results: List[BenchmarkResult]

File: agentbench-frontend-fixed/frontend/src/api/benchmark.ts
import apiClient from "./client";

/* ---------------- BENCHMARKS ---------------- */

export interface BenchmarkRequest {
  task_name: string;
  prompt: string;
  expected_output?: string;
  providers: string[];
}

export interface BenchmarkResult {
  provider: string;
  model: string | null;
  score: number;
  latency: number;
  tokens: number;
  cost: number;
  output: string;
  success: boolean;
  error?: string | null;
}

export interface BenchmarkResponse {
  benchmark_id: number;
  status: string;
  results: BenchmarkResult[];
}

export const benchmarkAPI = {
  // Runs synchronously and returns the full result set - the backend now
  // executes every selected provider (previously only the first selected
  // provider was ever run; the rest were silently ignored).
  runBenchmark: (data: BenchmarkRequest) =>
    apiClient.post<BenchmarkResponse>("/benchmarks/run", data),

  // Benchmark run history for the current user (backed by Task rows with
  // category "Benchmark" - see backend BenchmarkService.get_all).
  listBenchmarks: () => apiClient.get("/benchmarks"),

  // `benchmark_id` returned above is the underlying Task id, and
  // GET /benchmarks/{id}/results is scoped to the requesting user.
  getResults: (id: number) => apiClient.get(`/benchmarks/${id}/results`),
};
